osquery/logger/plugins/aws_util.cpp

```diff
@@ -179,10 +191,47 @@ OsqueryFlagsAWSCredentialsProvider::GetAWSCredentials() {
                                    FLAGS_aws_secret_access_key);
 }
 
-OsqueryAWSCredentialsProviderChain::OsqueryAWSCredentialsProviderChain()
+Aws::Auth::AWSCredentials
+OsquerySTSAWSCredentialsProvider::GetAWSCredentials() {
+  // Grab system time in seconds since epoch for token experation checks
+  auto now = std::chrono::system_clock::now();
+  auto epoch = now.time_since_epoch();
+  auto seconds = std::chrono::duration_cast<std::chrono::seconds>(epoch);
+  // Pull new STS creds if we dont have them cached from a previous run,
+  // otherwise return cached STS creds
+  if (token_expire_time <= seconds.count()) {
+    // Create and setup a STS client to pull our temp creds
+    VLOG(1) << "Generate new AWS STS credentials.";
+    initAwsSdk();
```

Comment:
Looking at it now, but I believe the `OsquerySTSAWSCredentialsProvider::GetAWSCredentials()` is a callback that gets triggered when auth fails since it was set int he `AWSCredentialsProviderChain`.

Playing around with some code to only call `initAwsSdk` on its first call.  I'll add the change if everything pans out.